### PR TITLE
[Public release 26/04] Introducing Mega MoE, FP4 Indexer and other features/fixes

### PR DESCRIPTION
**Pull request by LyricZhao**
### New features
- Mega MoE, fusing & overlapping dispatch/linear 1/SwiGLU/linear 2/combine into a single mega-kernel, overlapping NVLink communication and tensor core computation
  - Performance number: https://github.com/deepseek-ai/DeepGEMM/pull/316
  - Only FP8 x FP4 MoE is supported
  - Requires PyTorch >= 2.9
- FP4 Indexer (MQA logits) with larger MTP support
- FP8 x FP4 GEMM
- PDL
- Refactors on GEMM heuristics
- Faster JIT compilation
- GEMM optimizations (**dynamic** swap A/B, much faster MoE GEMM)
- DeepEPv2 MoE GEMM layout

### Bug fixes
- JIT may crash on distributed FS
- Some kernel hangs and IMA

### Contributors
- Mega MoE: @LyricZhao @zheanxu @bucket-xv @RayWang96 @interestingLSY @kurisu6912 @xay5421 @yukuai26 
- FP4 Indexer: @zheanxu @xay5421 @interestingLSY @kurisu6912
- GEMM, PDL, JIT and bug fixes: @zheanxu @bucket-xv @xay5421 @yukuai26 @LyricZhao

### Additional notes
Mega MoE is still under development and optimizations, stay tuned and optimization ideas are welcome!
**Disclaimer**: this release is only related to DeepGEMM's development, has nothing to do with internal model release.

### CODE REVIEW

**Review by zheanxu — APPROVED** [review]

**Comment by yuho8818** [comment]
LGTM

**Comment by huangzhilin-hzl** [comment]
Hi @LyricZhao  `tests/test_mega_moe.py` currently depends on `deep_ep.ElasticBuffer`, but I could not find `ElasticBuffer` in the public `deepseek-ai/DeepEP` branches; was this test written against a private/internal DeepEP version?

**Comment by LyricZhao** [comment]
@huangzhilin-hzl `deep_ep.ElasticBuffer` (DeepEP V2) will be released in this week. And, you can still skip the legacy path test and just look at the perf numbers.

**Comment by CyberQin** [comment]
deep_ep基础设施还没更新完啊，那model release不是更佳悬了，yifan的爆料要失真了

**Comment by qinqinwo** [comment]
请问有支持其他卡型的计划吗？比如H100。

**Comment by yy16432** [comment]
Any plan to add support to RTX PRO 6000(SM 120)?  Thanks!
